test: introduce `libp2p-swarm-test`

thomaseizinger opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->

This patch-set introduces `libp2p-swarm-test`. It provides utilities for quick and safe bootstrapping of tests for `NetworkBehaviour`s. The main design features are:

- Everything has timeouts
- APIs don't get in your way
- Minimal boilerplate

Closes #2884.

## Notes

I did not port `libp2p-kad` because the testing logic was too complicated. I also did not add tests for protocols that don't already have some. That is out of scope for this PR.

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

- ~Should this testing crate also pull together utilities like [`ArbitraryPeerId`](https://github.com/kpp/rust-libp2p/blob/bc77431620f952dd8a9e0054bba73bb1c9638bb7/protocols/kad/src/query/peers/closest/disjoint.rs#L535-L546)?~
	- No. That should go into `quickcheck-ext`.
- ~Can we make the `Swarm::new_ephemeral` API any better? At the moment it only allows to configure the underlying transport but the authentication and muxer are hardcoded. I don't want to introduce any more parameters because it will get too verbose and there wouldn't be a benefit in having the function in the first place. Should we perhaps introduce a dedicated `SwarmBuilder` for our tests that start with some reasonable test defaults, like a `MemoryTransport`?~
- ~Trying to port the new harness to `libp2p-ping`, I noticed that we'd lose the `MuxerChoice` as part of the prop test. If we go forward with the `SwarmBuilder` extension idea, we could make a function there that cycles through various muxers and authentication algorithms to make sure things work across various stacks. I am not sure it is worth it though to be honest.~

## Open tasks

- ~Port to more crates~

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
~- [ ] A changelog entry has been made in the appropriate crates~

thomaseizinger committed:
Initialise `libp2p-swarm-test` with bits of rendezvous' test-harness

thomaseizinger committed:
Introduce `SwarmExt::next_within`

Returns an event or panics in case nothing happens for X seconds.

thomaseizinger committed:
Rewrite libp2p-ping tests using new test framework

melekes reviewed (commented):
I'm in great support of this ❤️

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Only use yamux

thomaseizinger committed:
Introduce `SwarmExt::new_ephemeral`

thomaseizinger committed:
Add missing features

thomaseizinger committed:
Re-write first test of libp2p-ping with `SwarmExt::new_ephemeral`

This forces us to get rid of the muxer choice which I am not yet
sure about.

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,226 @@
+    /// Listens on a random memory address, polling the [`Swarm`] until the transport is ready to accept connections.
+    async fn listen_on_random_memory_address(&mut self) -> Multiaddr;
+
+    /// Listens on a random memory address, polling the [`Swarm`] until the transport is ready to accept connections.
```
```suggestion
    /// Listens on a random tcp address, polling the [`Swarm`] until the transport is ready to accept connections.
```

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,226 @@
+
+    async fn next_within(
+        &mut self,
+        seconds: u64,
```
Having to provide a timeout seems off to me. How would a user decide on the right value? Can't we just rely on the overall test timeout?

thomaseizinger replied:
Tests don't have an overall timeout unless the user wraps it all in one which is boilerplate and thus doesn't happen very often. Plus, compared to a test timeout, having the test failure in here provides a nice stacktrace that is easier to debug on where the test is failing.

The reason we can't have a timeout default is because we don't know about the test case, only the dev writing the test can decide that :)

mxinden replied:
> Tests don't have an overall timeout unless the user wraps it all in one which is boilerplate and thus doesn't happen very often. Plus, compared to a test timeout, having the test failure in here provides a nice stacktrace that is easier to debug on where the test is failing.

I wasn't aware of this. Thanks.

Still, I would consider concrete timeouts in tests a code smell. Users of this framework would have to make assumptions on the environment this is run on. E.g. they need to take into account that their test might run on a GitHub CI or some high performant developer machine.

I can not think of a way without the timeout while still having a strack trace. Is the strack trace important? I guess this boils down to "should one be able to see the root cause purely from the unit test failure, or do we expect a red test to need debugging anyways".

I don't expect any unit test to take longer than 10 seconds. In case we want to keep the stack traces, how about setting the timeout to a fixed `10s`?

thomaseizinger replied:
> I don't expect any unit test to take longer than 10 seconds. In case we want to keep the stack traces, how about setting the timeout to a fixed `10s`?

I guess for any situation where the timeout doesn't fit, the user can always fall back to the `StreamExt::next` function which doesn't have a timeout so this is probably a good idea?

What we need to balance with these test abstractions is for them to flexible enough to handle a lot of situations but also providing enough value in making test code concise. I think a timeout 10s is a good default and if it doesn't fit, you can always just code it yourself and it is not too verbose. If it happens a lot, we can look into adjusting the timeout or finding other solutions.

mxinden commented on swarm/src/lib.rs:
```diff
@@ -249,6 +249,16 @@ pub enum SwarmEvent<TBehaviourOutEvent, THandlerErr> {
 
+impl<TBehaviourOutEvent, THandlerErr> SwarmEvent<TBehaviourOutEvent, THandlerErr> {
+    /// Extract the `TBehaviourOutEvent` from this [`SwarmEvent`] in case it is the `Behaviour` variant, otherwise fail.
+    pub fn try_into_behaviour_event(self) -> Result<TBehaviourOutEvent, Self> {
```
```suggestion
    pub fn try_into_behaviour_event(self) -> Result<TBehaviourOutEvent, Self> {
```

Neat return signature!

thomaseizinger replied:
:pray:

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,226 @@
+        Self: Sized;
+
+    /// Establishes a connection to the given [`Swarm`], polling both of them until the connection is established.
+    async fn block_on_connection<T>(&mut self, other: &mut Swarm<T>)
```
```suggestion
    async fn connect<T>(&mut self, other: &mut Swarm<T>)
```

- `block_on` reminds me of `futures::executor::block_on` which this doesn't use.
- `block_on_connection` does not imply actually triggering the dial

What do you think of the new name?

thomaseizinger replied:
Yeah sounds good and it is shorter too :)

mxinden reviewed (commented):
Direction looks good to me. Thanks for pushing this forward.

thomaseizinger committed:
Rename connect function

thomaseizinger committed:
Fix typo

thomaseizinger committed:
Default transport in `new_ephemeral` to `MemoryTransport`

NB implementations should not make any assumptions about the
underlying transport implementations. Defaulting this to
`MemoryTransport` allows us to simplify the API a great deal.

thomaseizinger committed:
Update docs

thomaseizinger committed:
Make ping tests pass by fully converting to test harness

We remove the dynamic muxer choice because it is not the responsibility
of the ping tests to check that the muxers behave identically.

thomaseizinger committed:
Remove unused imports

thomaseizinger committed:
Apply clippy suggestion

thomaseizinger committed:
Remove extern crate statement

Not sure where that came from!

thomaseizinger committed:
Remove duplication in test

elenaf9 commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,183 @@
+        .parse::<Multiaddr>()
+        .unwrap();
+
+    addr
```
```suggestion
    Protocol::Memory(0).into()
```
The memory transport picks a random port if the port is 0.

thomaseizinger replied:
That is so much clearer, thank you!

elenaf9 commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,183 @@
+    /// The swarm will use a [`MemoryTransport`] together with a noise authentication layer and
+    /// yamux as the multiplexer. However, these details should not be relied upon by the test
+    /// and may change at any time.
+    fn new_ephemeral(behaviour_fn: impl FnOnce(PeerId, Keypair) -> Self::NB) -> Self
```
```suggestion
    fn new_ephemeral(behaviour_fn: impl FnOnce(Keypair) -> Self::NB) -> Self
```
Since the `PeerId` can be derived from the `Keypair`, what do you think about removing it as parameter?

thomaseizinger replied:
We could actually, yeah. It is a bit verbose though and many times, behaviours only need to know the `PeerId` and not the `Keypair`. Perhaps I can make some clever abstractions here so the user can choose ...

thomaseizinger replied:
Addressed in [`c6b670d` (#2888)](https://github.com/libp2p/rust-libp2p/pull/2888/commits/c6b670d19b4a4c27cab660bc12688aa6578dd69c).

elenaf9 commented on swarm/src/lib.rs:
```diff
@@ -249,6 +249,16 @@ pub enum SwarmEvent<TBehaviourOutEvent, THandlerErr> {
 
+impl<TBehaviourOutEvent, THandlerErr> SwarmEvent<TBehaviourOutEvent, THandlerErr> {
+    /// Extract the `TBehaviourOutEvent` from this [`SwarmEvent`] in case it is the `Behaviour` variant, otherwise fail.
+    pub fn try_into_behaviour_event(self) -> Result<TBehaviourOutEvent, Self> {
```
Is there a case where a user would call this method and not unwrap the result? If not, what do you think about panicking here instead of returning an Err?

thomaseizinger replied:
Given that this is a function in production code, I don't think panicking is a good option here.

elenaf9 replied:
Woops, I overlooked that it is in `swarm` and not `swarm-test`. Agree with you.

elenaf9 reviewed (approved):
Nice!

thomaseizinger committed:
Update swarm-test/src/lib.rs

Co-authored-by: Elena Frank <elena.frank@protonmail.com>

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Merge branch 'swarm-test' of github.com:libp2p/rust-libp2p into swarm-test

thomaseizinger committed:
Fix imports

thomaseizinger committed:
Fix compile error

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Remove configurable timeout from `next_within`

thomaseizinger committed:
Use all external addresses of swarm for dialling

thomaseizinger committed:
Add `SwarmExt::next_behaviour_event`

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Add license

thomaseizinger committed:
Fix build error

thomaseizinger committed:
Reduce fn arguments of `new_ephemeral` to `Keypair`

thomaseizinger committed:
Port tests of `request-response` protocol

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Shorten to `listen`

thomaseizinger committed:
Simplify combinators

thomaseizinger committed:
MemoryTransport can now handle `/p2p`

thomaseizinger committed:
Don't track address if we don't use it

thomaseizinger committed:
Use `SelectAll`

thomaseizinger committed:
Make `wait_for` an async fn

thomaseizinger committed:
Simplify `drain_poll`

thomaseizinger committed:
Migrate to `libp2p-swarm-test`

thomaseizinger committed:
Switch `libp2p-swarm-test` to `PlaintextConfig`

thomaseizinger committed:
Autogenerate event

thomaseizinger committed:
Migrate from `LocalPool` to `async` test

thomaseizinger committed:
Inline variable

thomaseizinger committed:
Don't use select! macro

thomaseizinger committed:
Use time-gated function for dialing

thomaseizinger committed:
Add `dial_and_wait`

thomaseizinger committed:
Add `SwarmExt::wait`

thomaseizinger committed:
Migrate dcutr to `libp2p-swarm-test`

thomaseizinger committed:
Fix clippy lints

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Address compile errors

thomaseizinger committed:
Don't depend on root crate in `swarm-test`

thomaseizinger committed:
Rewrite autonat client tests using swarm-test

thomaseizinger committed:
Rewrite autonat server tests using swarm-test

thomaseizinger committed:
Minor dcutr tweaks

thomaseizinger committed:
Remove harness in autonat

thomaseizinger committed:
Write mdns async-std tests using swarm-test

thomaseizinger committed:
Rewrite tokio tests

thomaseizinger committed:
Don't depend on async-std executor of swarm

thomaseizinger committed:
Make gossipsub tests compile

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Fix dcutr tests

thomaseizinger committed:
Fix gossipsub doc tests

thomaseizinger committed:
Fix ping tests

thomaseizinger committed:
Make autonat test more resilient

thomaseizinger committed:
Fix req-res test

thomaseizinger committed:
Fix test

thomaseizinger committed:
Rename `next_or_timeout` to `next_swarm_event`

The timeout is implied by the testing lib and this is more consistent
with `next_behaviour_event`.

thomaseizinger committed:
Get rid of rendezvous harness

thomaseizinger committed:
Use defaults of derive macro

thomaseizinger committed:
Merge branch 'master' into swarm-test

thomaseizinger committed:
Introduce `swarm_test::drive` to avoid starvation in rendezvous protocol

thomaseizinger committed:
Further improve drive

thomaseizinger commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,352 @@
+///
+/// Especially (2) is crucial for our usage of this function.
+/// If a [`Swarm`] is not polled, nothing within it makes progress.
+/// This can "starve" the other swarm which for example may wait for another message to be sent on a connection.
```
I spent a good two hours of debugging timeouts in the test and unwoken tasks until it dawned on me why things were not working. @elenaf9 I believe this might also be the reason for your timeouts in #3345.

This function may seem like a beast but after trying various solutions locally, I am convinced it makes sense to add this to the test-harness and use this instead of `future::join` for testing `Swarm`s.

thomaseizinger committed:
Fix clippy and docs

thomaseizinger committed:
Fix clippy

thomaseizinger committed:
Use `drive` in ping test

thomaseizinger committed:
Lazily compute vec length

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,353 @@
+use async_trait::async_trait;
```
Please add a license header here :)

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,353 @@
+/// For example, if you expect the first [`Swarm`] to emit 2 events, you should assign the first variable of the returned tuple value to an array of size 2.
+/// This works especially well if you directly pattern-match on the return value.
+///
+/// ## Type of event
+///
+/// This function utilizes the [`TryIntoOutput`] trait.
+/// Similar as to the number of expected events, the type of event is inferred based on your usage.
+/// If you match against a [`SwarmEvent`], the first [`SwarmEvent`] will be returned.
+/// If you match against your [`NetworkBehaviour::OutEvent`] type, [`SwarmEvent`]s which are not [`SwarmEvent::Behaviour`] will be skipped until the [`Swarm`] returns a behaviour event.
```
:exploding_head:

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,353 @@
+    let mut res1 = Vec::<Out1>::with_capacity(NUM_EVENTS_SWARM_1);
+    let mut res2 = Vec::<Out2>::with_capacity(NUM_EVENTS_SWARM_2);
+
+    while res1.len() < NUM_EVENTS_SWARM_1 || res2.len() < NUM_EVENTS_SWARM_2 {
+        match futures::future::select(swarm1.next_swarm_event(), swarm2.next_swarm_event()).await {
+            Either::Left((o1, _)) => {
+                if let Ok(o1) = o1.try_into_output() {
+                    res1.push(o1);
+                }
+            }
+            Either::Right((o2, _)) => {
+                if let Ok(o2) = o2.try_into_output() {
+                    res2.push(o2);
+                }
+            }
+        }
+    }
```
Long term it would be nice to detect missed wake-ups in a `NetworkBehaviour` implementation with `libp2p-swarm-test`. If I am not mistaken, that is not possible with the above `while` `select!` loop as long as the other swarm triggers a wake-up.

Lets defer to a future pull request.

thomaseizinger replied:
This is kind of a conflicting requirement. I added this because it is very easy to poll a `Swarm` in a wrong way in a test, thus it is often the _test_ that is buggy, not the `NetworkBehaviour` implementation.

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,353 @@
+
+    /// Create a new [`Swarm`] with an ephemeral identity.
+    ///
+    /// The swarm will use a [`MemoryTransport`] together with a noise authentication layer and
```
```suggestion
    /// The swarm will use a [`MemoryTransport`] together with [`Plaintext2Config`] authentication layer and
```

If I am not mistaken, this framework doesn't actually use noise.

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,353 @@
+            .await;
+
+        // Memory addresses are externally reachable because they all share the same memory-space.
+        self.add_external_address(memory_multiaddr.clone(), AddressScore::Infinite);
```
:+1: good idea

thomaseizinger replied:
Thanks! I ended up doing it this way because it allowed to create the `connect` abstraction which can then rely on retrieving the addresses via the external addresses.

mxinden commented on swarm-test/src/lib.rs:
```diff
@@ -0,0 +1,353 @@
+        let peer_id = PeerId::from(identity.public());
+
+        let transport = MemoryTransport::default()
+            .or_transport(libp2p_tcp::async_io::Transport::default())
```
Why do we need the TCP transport? Is this needed for `libp2p-autonat`?

thomaseizinger replied:
Yes, and `dcutr` needs it too.

mxinden commented on protocols/request-response/tests/ping.rs:
```diff
@@ -156,154 +160,127 @@ fn ping_protocol() {
+    swarm2.behaviour_mut().send_request(&peer1_id, ping.clone());
+
+    // Wait for swarm 1 to receive request by swarm 2.
+    let event = loop {
+        futures::select!(
+            event = swarm1.select_next_some() => {
+                if let SwarmEvent::Behaviour(request_response::Event::Message {
+                    peer,
+                    message: request_response::Message::Request { request, channel, .. }
+                }) = event {
+                    assert_eq!(&request, &ping);
+                    assert_eq!(&peer, &peer2_id);
+
+                    drop(channel);
+                    continue;
+                }
+            },
+            event = swarm2.select_next_some() => {
+                if let SwarmEvent::Behaviour(ev) = event {
+                    break ev;
+                }
+            },
+        )
+    };
```
Can this not be done via the new shiny `drive` function?

thomaseizinger replied:
Probably, I've introduced that one only at the very end!

thomaseizinger replied:
I am not sure actually, this one is a bit different because the two peers operate autonomously. `drive` is useful if you want to "step" through a test in a more linear fashion because it gives you access to the `Swarm`s inbetween. Here, we can just create to tasks that operate independently.

mxinden reviewed (approved):
Just looking at the diff, this is already a great step forward.

A couple of comments. Nothing major. Good to merge from my end.

thomaseizinger committed:
Update swarm-test/src/lib.rs

Co-authored-by: Max Inden <mail@max-inden.de>

thomaseizinger committed:
Add license header

thomaseizinger committed:
Fix docs

mergify[bot] committed:
Merge branch 'master' into swarm-test
